metriks/gauge.go

```diff
@@ -0,0 +1,88 @@
+package metriks
+
+import (
+	"context"
+	"sync/atomic"
+	"time"
+
+	"github.com/armon/go-metrics"
+)
+
+const (
+	defaultGaugeDuration = time.Second * 10
+)
+
+// PersistentGauge will report on an interval the value to the metrics collector.
+// Every call to the methods to modify the value immediately report, but if we
+// don't have a change inside the window (default 10s) after the last report
+// we will report the current value.
+type PersistentGauge struct {
+	name  string
+	value int32
+	tags  []metrics.Label
+
+	ticker *time.Ticker
+	cancel context.CancelFunc
+	dur    time.Duration
+}
+
+// Set will replace the value with a new one, it returns the old value
+func (g *PersistentGauge) Set(value int32) int32 {
+	v := atomic.SwapInt32(&g.value, value)
+	g.report(value)
+	return v
+}
+
+// Inc will +1 to the current value and return the new value
+func (g *PersistentGauge) Inc() int32 {
+	v := atomic.AddInt32(&g.value, 1)
+	g.report(v)
+	return v
+}
+
+// Dec will -1 to the current value and return the new value
+func (g *PersistentGauge) Dec() int32 {
+	v := atomic.AddInt32(&g.value, -1)
+	g.report(v)
+	return v
+}
+
+func (g *PersistentGauge) report(v int32) {
+	Gauge(g.name, float32(v), g.tags...)
+	g.ticker.Reset(g.dur)
+}
+
+func (g *PersistentGauge) start(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-g.ticker.C:
+			g.report(g.value)
+		}
+	}
+}
+
+// Stop will make the gauge stop reporting. Any calls to Inc/Set/Dec will still report
+// to the metrics collector
+func (g *PersistentGauge) Stop() {
+	g.cancel()
+}
+
+// NewGauge will create and start a PersistentGauge that reports the current value every 10s
+func NewGauge(name string, tags ...metrics.Label) *PersistentGauge {
+	return NewGaugeWithDuration(name, tags, defaultGaugeDuration)
+}
+
+func NewGaugeWithDuration(name string, tags []metrics.Label, dur time.Duration) *PersistentGauge {
+	ctx, cancel := context.WithCancel(context.Background())
+	g := PersistentGauge{
+		name:   name,
+		tags:   tags,
+		ticker: time.NewTicker(dur),
+		cancel: cancel,
+		dur:    dur,
+	}
```

Comment:
What do you think about having this take a callback function that computes the value at the time of report? A user could register this with a function that reads the gauge value (eg. current length of queue) and we'd invoke it once every 10secs and report as a gauge. That could ensure you always have a reading every interval.

Reply:
Interesting, I kinda like it. But one thing that I'd worry about is that this value is actually being buffered outside of us by the agent. So if we're reporting every 10s, that doesn't mean that we're in sync with that windowing. The way that this is done we report every value to the agent (so high throughput ones would be caught there). We'd also lose the Inc/Dec features. 

I am thinking I'll make another gauge 'ScheduledGauge' that will do this. Seems really useful